### PR TITLE
Provide API to support context run configuration in Kotlin multiplatform projects

### PR DESCRIPTION
**Pull request by pTalanov**

### CODE REVIEW

File: platform/lang-api/src/com/intellij/execution/actions/MultipleRunLocationsProvider.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+package com.intellij.execution.actions
+
+import com.intellij.execution.Location
+import com.intellij.openapi.extensions.ExtensionPointName
+import com.intellij.openapi.extensions.Extensions
+
+abstract class MultipleRunLocationsProvider {
```
Comment (yole) [inline platform/lang-api/src/com/intellij/execution/actions/MultipleRunLocationsProvider.kt]: Please provide javadocs for all new public APIs

File: platform/lang-api/src/com/intellij/execution/actions/ConfigurationFromContext.java [hunk 2]
```diff
@@ -31,6 +32,12 @@
  * @see RunConfigurationProducer
  */
 public abstract class ConfigurationFromContext {
+
+  private boolean isFromAlternativeLocation;
+
+  @Nullable
+  private String alternativeLocationDisplayName;
```
Comment (yole) [inline platform/lang-api/src/com/intellij/execution/actions/ConfigurationFromContext.java]: Please use "my" prefix for fields in Java code

File: platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java [hunk 3]
```diff
@@ -140,4 +157,38 @@ private static ConfigurationFromContext findConfigurationFromContext(final Locat
...
+    SmartList<ConfigurationFromContext> result = new SmartList<>();
+    for (Location alternativeLocation : alternativeLocationsInfo.getAlternativeLocations()) {
+      ConfigurationContext fakeContextForAlternativeLocation = new ConfigurationContext(alternativeLocation);
+      List<ConfigurationFromContext> configurationsForLocation =
+        doGetConfigurationsFromContext(alternativeLocation, fakeContextForAlternativeLocation, strict);
+      if (configurationsForLocation != null) {
+        for (ConfigurationFromContext configurationFromContext : configurationsForLocation) {
+          configurationFromContext.setFromAlternativeLocation(true);
+          String locationDisplayName = alternativeLocationsInfo.getProvider().getLocationDisplayName(alternativeLocation, originalLocation);
+          configurationFromContext.setAlternativeLocationDisplayName(
```
Comment (yole) [inline platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java]: No need for line wrapping in this call

File: platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java [hunk 4]
```diff
@@ -140,4 +157,38 @@ private static ConfigurationFromContext findConfigurationFromContext(final Locat
...
+            locationDisplayName
+          );
+          RunConfiguration configuration = configurationFromContext.getConfiguration();
+          if (configuration instanceof LocatableConfigurationBase && ((LocatableConfiguration)configuration).isGeneratedName()) {
+            configuration.setName(configuration.getName() + " " + locationDisplayName);
+            ((LocatableConfigurationBase)configuration).setNameChangedByUser(true);
```
Comment (yole) [inline platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java]: We need to handle renames of classes/methods from which these configurations have been created

File: platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java [hunk 5]
```diff
@@ -140,4 +157,38 @@ private static ConfigurationFromContext findConfigurationFromContext(final Locat
     }
     return null;
   }
+
+  @Nullable
+  private static List<ConfigurationFromContext> getConfigurationsFromAlternativeLocations(
```
Comment (yole) [inline platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java]: Some tests for this logic would be nice

File: platform/lang-impl/src/com/intellij/execution/actions/BaseRunConfigurationAction.java [hunk 6]
```diff
@@ -65,14 +68,16 @@ protected BaseRunConfigurationAction(final String text, final String description
     final ConfigurationContext context = ConfigurationContext.getFromContext(dataContext);
     if (Registry.is("suggest.all.run.configurations.from.context") || context.findExisting() == null) {
       final List<ConfigurationFromContext> producers = getConfigurationsFromContext(context);
+      boolean isMultipleConfigurationsFromAlternativeLocations = producers.size() > 1 && producers.get(0).isFromAlternativeLocation();
+      boolean isRunAction = this instanceof RunContextAction && ((RunContextAction)this).isForExecutor(DefaultRunExecutor.EXECUTOR_ID);
+      boolean createRunAll = isMultipleConfigurationsFromAlternativeLocations && isRunAction;
       if (producers.size() > 1) {
-        final AnAction[] children = new AnAction[producers.size()];
+        final AnAction[] children = new AnAction[producers.size() + (createRunAll ? 1 : 0)];
```
Comment (yole) [inline platform/lang-impl/src/com/intellij/execution/actions/BaseRunConfigurationAction.java]: I think this would be easier with an `ArrayList`

**Comment by Vassiliy-Kudryashov** [comment]
Pavel, the code looks mostly harmless, but Dmitry's comments are very good TODOs.

**Comment by chashnikov** [comment]
Merged to master, thank you for contribution!

**Pull request closed without merging**